httpx/content_streams.py

```diff
@@ -68,10 +76,48 @@ def get_headers(self) -> typing.Dict[str, str]:
         content_length = str(len(self.body))
         return {"Content-Length": content_length}
 
+    def __iter__(self) -> typing.Iterator[bytes]:
+        yield self.body
+
     async def __aiter__(self) -> typing.AsyncIterator[bytes]:
         yield self.body
 
 
+class IteratorStream(ContentStream):
+    """
+    Request content encoded as plain bytes, using an byte iterator.
+    """
+
+    def __init__(
+        self, iterator: typing.Iterator[bytes], close_func: typing.Callable = None
+    ) -> None:
+        self.iterator = iterator
+        self.close_func = close_func
+        self.is_stream_consumed = False
+
+    def can_replay(self) -> bool:
+        return False
+
+    def get_headers(self) -> typing.Dict[str, str]:
+        return {"Transfer-Encoding": "chunked"}
+
+    def __iter__(self) -> typing.Iterator[bytes]:
+        if self.is_stream_consumed:
+            raise StreamConsumed()
+        self.is_stream_consumed = True
+        for part in self.iterator:
+            yield part
+
+    async def __aiter__(self) -> typing.AsyncIterator[bytes]:
+        raise RuntimeError("Attempted to call a async iterator on an sync stream.")
+        async for part in ():  # pragma: nocover
+            yield b""
```

Comment:
Does this have any actual runtime consequence, or is this to please linters?

Turns out that we can make this work without tricks by dropping the `async` annotation…

```python
def __aiter__(self) -> typing.AsyncIterator[bytes]:
    raise RuntimeError("Attempted to call a async iterator on an sync stream.")
```

Reply:
Nice one, yup!